### ISSUE TITLE
Matrix alignment and row spacing problems of mathematical models in tables

### ISSUE THREAD

**Issue by whateverzpy** [typst/typst#5489]
### Description

Look at the code and the image below
```typst
#table(
  columns: (
    auto,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
  ),align: horizon + center,rows: (auto, 3em, 3em, 3em),stroke: none,
  table.hline(),
  [$k$],
  table.vline(stroke: 0.6pt),
  [$bold(x)^((k))$],
  table.vline(stroke: 0.6pt),
  [$nabla f(bold(x)^((k)))$],
  table.vline(stroke: 0.6pt),
  [$bold(H)^((k))$],
  table.vline(stroke: 0.6pt),
  [$bold(d)^((k))$],
  table.vline(stroke: 0.6pt),
  [$lambda_k$],
  table.vline(stroke: 0.6pt),
  [$bold(s)^((k))$],
  table.vline(stroke: 0.6pt),
  [$bold(y)^((k))$],
  table.hline(stroke: 0.6pt),
  [$1$],[$mat(0;0)$],[$mat(-4;-8)$],[$mat(1,0;0,1)$],[$mat(4;8)$],[$5 / 34$],[$mat(10/17;20/17)$],[$mat(20/17;160/17)$],
  table.hline(stroke: 0.6pt),
  [$2$],[$mat(10/17;20/17)$],[$mat(-48/17;24/17)$],[$mat(2241/2210,-71/1105;-71/1105,147/1105)$],[$mat(192/65;-24/65)$],[$65 / 136$],[$mat(24/17;-3/17)$],[$mat(48/17;-24/17)$],
  table.hline(stroke: 0.6pt),
  [$3$],[$mat(2;1)$],[$mat(0;0)$],[],[],[],[],[],
  table.hline()
)
```
![image](https://github.com/user-attachments/assets/7baadbbd-7071-488d-95a8-23e4d733aa1c)
This matrix is obviously not aligned with the other matrices in this row. 
However, when I don't specify line spacing, it seems like...
```typst
#table(
  columns: (
    auto,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
    0.125fr,
  ),align: horizon + center,stroke: none,
  table.hline(),
  [$k$],
  table.vline(stroke: 0.6pt),
  [$bold(x)^((k))$],
  table.vline(stroke: 0.6pt),
  [$nabla f(bold(x)^((k)))$],
  table.vline(stroke: 0.6pt),
  [$bold(H)^((k))$],
  table.vline(stroke: 0.6pt),
  [$bold(d)^((k))$],
  table.vline(stroke: 0.6pt),
  [$lambda_k$],
  table.vline(stroke: 0.6pt),
  [$bold(s)^((k))$],
  table.vline(stroke: 0.6pt),
  [$bold(y)^((k))$],
  table.hline(stroke: 0.6pt),
  [$1$],[$mat(0;0)$],[$mat(-4;-8)$],[$mat(1,0;0,1)$],[$mat(4;8)$],[$5 / 34$],[$mat(10/17;20/17)$],[$mat(20/17;160/17)$],
  table.hline(stroke: 0.6pt),
  [$2$],[$mat(10/17;20/17)$],[$mat(-48/17;24/17)$],[$mat(2241/2210,-71/1105;-71/1105,147/1105)$],[$mat(192/65;-24/65)$],[$65 / 136$],[$mat(24/17;-3/17)$],[$mat(48/17;-24/17)$],
  table.hline(stroke: 0.6pt),
  [$3$],[$mat(2;1)$],[$mat(0;0)$],[],[],[],[],[],
  table.hline()
)
```
![image](https://github.com/user-attachments/assets/fa0ed1b3-df0d-4567-9880-a04614ef7c44)
Multiple matrices will collide directly with the table edge line.

### Operating system

Windows

### Typst version

- [X] I am using the latest version of Typst

**Comment by whateverzpy** [comment]
The reason for the seeming misalignment may have to do with the column width not being given enough.
70pt
![image](https://github.com/user-attachments/assets/1848fbb4-f497-4ed6-89a4-80c211e9942a)
60pt
![image](https://github.com/user-attachments/assets/38630192-e3a1-4c9e-9216-273326d8cf1b)

**Comment by laurmaedje** [comment]
Minimized:
```typ
#table(
  columns: (
    2cm,
    2cm,
  ),
  $mat(1/1;2/2)$,
  $mat(3/3, 4/4; 5/123456, 1/123)$
)
```
Removing just one letter from "123456" or "123" changes things. Very odd.

**Comment by mkorje** [comment]
I've figured out the cause, it's due to the `linebreak_simple` function in `inline/linebreak.rs`.

This is the example I'll use:
```typ
--- playground ---
#set page(width: auto, height: auto)
#table(
  columns: (70pt,) * 1,
  align: horizon + center,
  stroke: 0.6pt,
  [$mat(2241/2210,-71/1105;-71/1105,147/1105)$],
)
#table(
  columns: (60pt,) * 1,
  align: horizon + center,
  stroke: 0.6pt,
  [$mat(2241/2210,-71/1105;-71/1105,147/1105)$],
)
```
It gets laid out inline and after preparation the items are as follows:
```
[(0..3, Skip("\u{2066}")), (3..6, Frame(Frame ... )), (6..9, Skip("\u{2069}"))]
```
When it gets to the `linebreak_simple` function, the first table has width argument `60.0pt` and the second has `50.0pt`.
Both have two breakpoints calculated: one Normal at end 3, and a Mandatory at end 9.
A (blank) line is created and stored in last for the first item (0..3). But then, when the second line gets created, it does not fit in the case of the second table, which causes the blank line to be pushed. Whereas with the first table, it does fit and so the blank line is replaced.

TLDR: It doesn't seem like there should be a linebreak opportunity immediately after the Left-To-Right Isolate (\u{2066}). Is this then an upstream bug in `icu_segmenter::LineSegmenter` maybe?

**Comment by laurmaedje** [comment]
> It doesn't seem like there should be a linebreak opportunity immediately after a Left-To-Right Isolate (\u{2066}). Is this then an upstream bug in icu_segmenter::LineSegmenter maybe?

Could indeed be a bug.

**Comment by mkorje** [comment]
I looked into this last week and I don't think it's a bug in LineSegmenter. (Completely forgot to actually reply here, sorry about that!)

Checking the C library, it gives the same boundary positions:
<details>

<summary>icu77 in C</summary>

```c
#include <stdlib.h>
#include <stdio.h>
#include "unicode/ustring.h"
#include "unicode/ubrk.h"

int main(void) {
    // Left-To-Right Isolate, Object Replacement Character, Pop Directional Isolate
    const UChar text[]={ 0x2066, 0xFFFC, 0x2069, 0 };
    int32_t len = -1;

    UBreakIterator* bi;
    int32_t p;
    UErrorCode err = U_ZERO_ERROR;
    bi = ubrk_open(UBRK_LINE, 0, text, len, &err);
    if (U_FAILURE(err)) return EXIT_FAILURE;
    p = ubrk_first(bi);
    while (p != UBRK_DONE) {
        printf("Boundary at position %d\n", p);
        p = ubrk_next(bi);
    }
    ubrk_close(bi);
  
    return EXIT_SUCCESS;
}

// Output:
// Boundary at position 0
// Boundary at position 1
// Boundary at position 3
```

</details>

Back to Rust then, the following shows the kind of line break each of these characters give:
```rust
use icu::properties::{LineBreak, maps};
use icu::segmenter::LineSegmenter;

fn main() {
    assert_eq!(maps::line_break().get('\u{2066}'), LineBreak::CombiningMark);
    assert_eq!(
        maps::line_break().get('\u{FFFC}'),
        LineBreak::ContingentBreak
    );
    assert_eq!(maps::line_break().get('\u{2069}'), LineBreak::CombiningMark);

    let segmenter = LineSegmenter::new_auto();
    let breakpoints: Vec<usize> = segmenter.segment_str("\u{2066}\u{FFFC}\u{2069}").collect();
    // Intended, see LB10 (https://www.unicode.org/reports/tr14/#LB10)
    assert_eq!(&breakpoints, &[0, 3, 9]);
}
```
As mentioned [here](https://www.unicode.org/reports/tr14/#CM), most control and formatting characters are given class CM (combing mark), including the two isolate characters in the example.

I think the following is what leads to the observed behaviour.
+ [LB9](https://www.unicode.org/reports/tr14/#LB9): "\u{FFFC}\u{2069}" is treated as just '\u{FFFC}', which has class CB (contingent break opportunity). This by default has a break opportunity both [before and after](https://www.unicode.org/reports/tr14/#CB).
+ [LB10](https://www.unicode.org/reports/tr14/#LB10): The remaining '\u{2066}' at the start is treated like a capital latin A.

**Comment by saecki** [comment]
> As mentioned [here](https://www.unicode.org/reports/tr14/#CM), most control and formatting characters are given class CM (combing mark), including the two isolate characters in the example.
> 
> I think the following is then what leads to the observed behaviour.
> 
>     * [LB9](https://www.unicode.org/reports/tr14/#LB9): "\u{FFFC}\u{2069}" is treated as just '\u{FFFC}', which has class CB (contingent break opportunity). This by default has a break opportunity both [before and after](https://www.unicode.org/reports/tr14/#CB).
> 
>     * [LB10](https://www.unicode.org/reports/tr14/#LB10): The remaining '\u{2066}' at the start is treated like a capital latin 

Thanks for looking in to this! This is what I've come to understand as well.

I'm currently working on hopefully more correct line breaking by looking at the character before and after the break opportunity.
So in this case the combining mark could be ignored (categorized as `AL`, and the break opportunity of the `OBJECT REPLACEMENT CHARACTER` (`\u{FFFC}`) could be handled by us in some more meaningful way?

Would be great if the `LineSegmenter` would provide more context than just the index 😅

**Comment by saecki** [comment]
I've tried a bunch of things, and in the end this seemed like the best option. If anyone (especially @mkorje) has some input on if this is a good idea or not that would be appreciated.
